### PR TITLE
perf(encode): close a third of the level-1 gap to libzstd

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Level 1 on real data was the widest gap on the encoder, and #493 filed it as an
entropy-stage problem. Part of it was. The largest single cost turned out to be
somewhere the issue did not look: every window slide cleared the hash table and
rehashed the whole retained tail — a pass over every byte the window kept, once
per `max_window_size` bytes.

On the 8 MiB access log at level 1 the CLI goes from **1.82x libzstd to 1.42x**;
`--fast=1` from 1.95x to 1.50x; 2 MiB of incompressible input from 1.52x to
**0.57x**, which we now win. In the loop harness at level 1: 6.19 G cycles ->
4.72 G against libzstd's 2.91 G, so **2.13x -> 1.62x**.

## What landed, biggest first

**The window slide rebuilt the table.** Upstream's `ZSTD_reduceIndex` subtracts
the correction from each stored position; a pass over the table rather than
over the window's bytes. Level 1 on the access log: **-11.6% cycles, -18.8%
instructions**. It also leaves the table holding what it held going in — the
rehash indexed positions the matcher's step had skipped and never stored.

**The dense prime loop was a `RangeInclusive`.** A call-graph profile put 12% of
a level-1 encode inside its `next` and `lt` — the exhausted flag, live across a
loop whose body is one hash and one store. The comment above it already
recorded that `step_by(1)` had cost a factor of two there for the same reason;
the inclusive range it was replaced with carried the other half. Half-open
range: **-7.8% cycles, -14.3% instructions** at level 1, and **-42% instructions**
on incompressible input, where the prime was most of the frame.

**The second sequence array is gone.** The emitter copied every sequence into a
second array for the sake of one field. The code now goes into the slot the
found offset occupied, which has no reader once its code exists — upstream's
single `SeqDef::offBase`. -2.4% at level 1.

**Deriving the offset codes and counting them are one pass**, not two walks of
the same array where the second read back what the first wrote. -3.3% at level 1.

**The repeat-offset history stays in registers** across that pass. Behind the
caller's reference it was three stores into the compressor per sequence,
because the loop also writes through the sequence slice. -2.5%.

**The codes cross to the writer instead of being re-derived**, packed into one
word: two codes at six bits, the offset code at five, two extra-bit widths at
five. This is #493's lead 2, and upstream's `ZSTD_seqToCodes` shape. -0.2 to
-3.4% across the grid, neutral at level 1 where the array traffic is largest.

**The unchecked bit add lost its zero-width branch** — six calls a sequence in
a writer that is a quarter of the frame. A masked shift says the same thing and
costs nothing to say. -1.2% at level 1, instructions down everywhere.

**The highest sequence code comes from one scan a block**, not three compares
a sequence. Each histogram is scanned from its own alphabet's last code (35,
52, 31), as upstream's `HIST_count_simple` scans it. -1.4% at level 1. A block
of sixteen sequences or fewer, the size the post-split and a small frame
produce, takes the maxima from the codes it just wrote instead, which retires
0.3 to 0.5% fewer instructions on those encodes than the scans do.

**`HuffNode` is 12 bytes, not 40.** Four `usize` fields plus an `Option` made
the node table twenty kilobytes for a full alphabet, walked several times a
block beside a six-kilobyte histogram; upstream's `nodeElt` is eight bytes.
Every field is bounded by the block, and the entry points now say so: they are
public, so a histogram whose counts do not fit a node is refused before
anything reads it rather than truncated into a leaf. Ahead of everything,
because the cheap path's table-log pick needs the total, and summing it in a
`usize` overflows a 32-bit target before the tree is ever built; the check
hands its total on, so the cheap path walks the histogram once. Kept on its
instruction count — the clock was ambiguous.

## Measurement

i9, prebuilt arms alternating in one session, `perf stat -r 5`, two passes,
against zstd 1.5.7 on the same machine. The grid is every level from
`--fast=5` to 19 against five fixture shapes — decodecorpus with and without a
dictionary, an 8 MiB access log, 2 MiB of incompressible bytes, and a 10 KiB
random block with a dictionary — 110 rows in all. Median **-0.81%** cycles,
87 rows of 110 negative. A representative slice:

| fixture | level | bytes | vs main | vs C | cycles main | cycles new | delta | new/C |
|---|---|---|---|---|---|---|---|---|
| access log 8 MiB | --fast=1 | 1,924,368 | -0.090% | -0.18% | 158,246,486 | 121,008,420 | **-23.53%** | 1.49x |
| access log | 1 | 1,467,858 | +0.094% | -0.07% | 159,269,174 | 123,852,670 | **-22.24%** | 1.41x |
| access log | 3 | 1,473,262 | same | -0.16% | 203,134,934 | 196,205,708 | -3.41% | 1.65x |
| access log | 5 | 1,335,833 | same | -0.13% | 371,145,046 | 365,524,740 | -1.51% | 1.59x |
| access log | 9 | 1,162,003 | same | +0.29% | 706,053,876 | 697,090,930 | -1.27% | 1.35x |
| access log | 13 | 1,076,287 | same | +0.05% | 1,756,598,678 | 1,760,676,816 | +0.23% | 1.05x |
| access log | 19 | 978,710 | same | -0.02% | 14,141,942,304 | 14,104,883,790 | -0.26% | **0.97x** |
| decodecorpus z000033 | --fast=1 | 595,177 | same | -0.05% | 14,529,101 | 14,096,568 | -2.98% | 1.30x |
| z000033 | 1 | 571,128 | same | -0.07% | 17,307,497 | 16,845,344 | -2.67% | 1.29x |
| z000033 | 4 | 496,150 | same | -1.86% | 51,696,853 | 51,896,979 | +0.39% | 1.93x |
| z000033 | 5 | 486,821 | same | -1.77% | 64,300,286 | 63,249,850 | -1.63% | 1.43x |
| z000033 | 9 | 484,012 | same | -1.49% | 88,234,092 | 86,907,795 | -1.50% | 1.24x |
| z000033 | 13 | 482,502 | same | -1.49% | 138,196,198 | 136,375,388 | -1.32% | **0.94x** |
| z000033 | 19 | 426,362 | same | -0.04% | 737,030,732 | 730,801,596 | -0.85% | 1.26x |
| z000033 + 16 KiB dict | 1 | 551,588 | same | -0.05% | 18,292,121 | 17,726,262 | -3.09% | 1.16x |
| z000033 + dict | 5 | 468,587 | same | +0.99% | 67,438,038 | 66,334,156 | -1.64% | **0.79x** |
| z000033 + dict | 9 | 464,143 | same | -1.09% | 114,707,604 | 110,936,949 | -3.29% | 1.31x |
| z000033 + dict | 16 | 422,218 | same | -0.18% | 645,883,126 | 640,568,704 | -0.82% | 1.26x |
| incompressible 2 MiB | 1 | 2,097,217 | same | +0.00% | 9,526,587 | 3,441,284 | **-63.88%** | **0.58x** |
| incompressible | 9 | 2,097,217 | same | +0.00% | 6,224,291 | 6,264,472 | +0.65% | **0.44x** |
| incompressible | 19 | 2,097,217 | same | +0.00% | 13,566,243 | 13,539,748 | -0.20% | **0.02x** |
| 10 KiB random + dict | 1 | 10,259 | same | +0.01% | 734,935 | 731,866 | -0.42% | 1.05x |
| 10 KiB random + dict | 9 | 10,259 | same | +0.01% | 1,028,057 | 989,294 | -3.77% | 1.06x |
| 10 KiB random + dict | 19 | 10,259 | same | +0.01% | 2,733,174 | 2,731,065 | -0.08% | 1.18x |

Dictionary rows through the harness that prepares a CDict once and reuses the
context, so no process start is in the measurement:

| fixture | level | main | new | delta | libzstd | new/C |
|---|---|---|---|---|---|---|
| 4 KiB log lines + dict | 1 | 209,175,294 | 208,102,933 | -0.51% | 117,743,508 | 1.77x |
| 4 KiB log lines + dict | 5 | 836,786,127 | 783,052,199 | **-6.42%** | 538,738,221 | 1.45x |
| 4 KiB log lines + dict | 11 | 799,219,916 | 793,692,540 | -0.69% | 525,847,159 | 1.51x |
| 10 KiB random + dict | 13 | 7,688,019,770 | 7,682,941,157 | -0.07% | 5,980,848,160 | 1.28x |
| 10 KiB random + dict | 16 | 7,643,976,091 | 7,640,597,157 | -0.04% | 5,985,196,693 | 1.28x |

## Output

Byte-identical on 105 of the 110 rows. The five that move are the Fast band of
the access log, the only fixture here long enough for the window to slide: the
change moves them in both directions and by under a seventh of a percent, and
they stay smaller than libzstd's. `--fast=5` 2,298,225 -> 2,297,015 bytes,
`--fast=3` 2,131,861 -> 2,129,095, `--fast=1` 1,926,093 -> 1,924,368, level 1
1,466,481 -> 1,467,858, level 2 1,515,569 -> 1,516,103, against libzstd's
2,300,531 / 2,133,375 / 1,927,913 / 1,468,937 / 1,517,054. Level 3 and above are
a different backend and unchanged, and so is every other fixture at every
level.

## Four dead ends, measured, with their numbers in the commits

- A fourth field for the wire code widened the sequence from twelve bytes to
  sixteen: **+10.3% cycles** at level 5, instructions flat. The copy was never
  the cost, the width was.
- Four interleaved count tables, upstream's `HIST_count_parallel_wksp` shape:
  +0.2 to +2.3% instructions, +1.1 to +2.2% cycles. The lane index and the fold
  cost more than the store-forwarding chain they break.
- Splitting the symbol out of `encode_literal_length` so the histogram skips
  the extra-bit lookup: the two bounds checks do leave the disassembly and it
  still measured +0.16 to +1.71% instructions.
- `inline(always)` on those split helpers, which changed the instruction count
  by two in fifteen billion — they were already inlined and the call-overhead
  explanation was wrong.

## On the rows that moved without work

Several rows move by one to four percent with retired instructions flat or
DOWN, in both directions, and the direction depends on which binary the row
was measured in. The widest one in this grid is the access log at level 10,
which the CLI puts at +4.3% and the loop harness at -2.0% — the same row, the
same input, instructions down in both (2,039.1 M -> 2,033.9 M in the CLI,
16,039.5 M -> 15,972.6 M in the loop). That is code layout, a property of the
binary rather than of the change. Read the instruction column on those rows.

## What is left in #493

Lead 1's second half — the Fast kernel emitting the offBase it already knows.
Measured at **3.18% of the level-1 frame**, and it is a serial dependency
rather than instruction count, so only moving the decision into the kernel
removes it. It needs the kernel's repeat-offset stack to become rollback-able
at the three encoder discard sites first, which is #500.

## Related

- #505: explicit compression parameters do not reach a dictionary frame's
  tables, so no configuration today puts a Fast table far above its window.
  Fixing it makes that shape reachable, and the index slide gets re-measured
  there as part of it.

## Testing

- `cargo nextest run -p structured-zstd -F hash,std,dict-builder`: 1059 passed
- `cargo nextest run -p ffi-bench -F bench-internals,dict-builder`: 64 passed
- `cargo test --doc`: 23 passed
- output compared against main over the whole 110-row grid: five fixture shapes
  against every level from `--fast=5` to 19
- `cargo clippy --all-targets` on both CI feature sets, on
  `--no-default-features --features kernel-scalar`, and the wasm32 target;
  `cargo fmt --check`: clean
- output compared over 60 rows after every commit: three fixture shapes against
  ten levels, each run both plain and dictionary-primed

Part of #493.

## Summary by CodeRabbit

- **Performance Improvements**
  - Improved compression efficiency and reduced temporary processing overhead.
  - Optimized fast matching as data windows advance while preserving useful match information.
  - Reduced memory usage in Huffman encoding through more compact representations.

- **Reliability**
  - Improved handling of full bit accumulators and zero-length bit writes.
  - Added validation for oversized Huffman inputs while preserving maximum-size valid inputs.

- **Tests**
  - Expanded coverage for sequence handling, boundary conditions, estimator consistency, memory accounting, and Huffman limits.

- **Tools**
  - Added an example for benchmarking compression with oversized hash tables.

### CODE REVIEW

File: zstd/src/encoding/blocks/compressed.rs [hunk 1]
```diff
@@ -141,10 +141,14 @@ struct EncodedBlockParts {
 #[derive(Default)]
 pub(crate) struct CompressedBlockScratch {
     parts: EncodedBlockParts,
+    /// One packed [`SequenceCode`] per sequence of the partition being
+    /// encoded, filled by the pass that derives the offset codes and read by
+    /// the bit writer. Kept here so it is allocated once for the compressor
+    /// rather than per block.
+    sequence_codes: Vec<u32>,
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/blocks/compressed.rs:148]: @coderabbitai Fixed — `sequence_codes.capacity() * size_of::<u32>()` is now in `retained_heap_size`. It is a retained buffer like every other one in the scratch, and leaving it out under-reported the context through `ZSTD_sizeof_CCtx` by four bytes per sequence of retained capacity. Carries a regression test (`retained_heap_size_counts_the_sequence_code_buffer`) that reserves into the buffer and asserts the reported size grows by at least the reserved bytes; it fails on the previous commit.

File: zstd/src/encoding/blocks/compressed/tests.rs [hunk 2]
```diff
@@ -383,13 +383,12 @@ fn estimator_literals_section_mirrors_emit_for_short_inputs() {
...
             let mut emitted: Vec<u8> = Vec::new();
-            let mut scratch: Vec<crate::blocks::sequence_section::Sequence> = Vec::new();
-            encode_block_parts_with_sequence_scratch(
+            encode_block_parts(
                 &mut emit_state,
                 &literals,
-                &[],
+                &mut [],
+                &mut Vec::new(),
                 &mut emitted,
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/blocks/compressed/tests.rs:391]: @coderabbitai Added `estimator_and_emitter_agree_on_a_block_with_sequences` — 64 sequences with repeats among the offsets, run through both paths on Fast, Lazy and BtUltra2.

Worth recording what it found, because it changed what the test asserts. Exact equality does NOT hold on the sequences section, and is not meant to: `fse_bit_cost` / `cross_entropy_cost` price a symbol at its average width from the normalised probability — the same model as upstream's `ZSTD_fseBitCost` — while the writer pays what the state trajectory actually costs. The divergence is a stable two bytes on this fixture across all three strategies. Exactness holds only on the literals section, where Huffman code lengths are known per symbol, which is why the existing parity tests pass empty sequence arrays.

So the test asserts the contract that does hold: both sides come out with the same repeat-offset history (a wrong offset code on either side moves the histogram, the table choice and the price by far more than rounding), and the price stays within the model's rounding, bounded at four bytes plus one per sixteen sequences. The reasoning is in the test's own doc comment.

File: zstd/src/huff0/huff0_encoder.rs [hunk 3]
```diff
@@ -1230,9 +1243,9 @@ fn build_huffman_leaf_depths_into(counts: &[usize], nodes: &mut Vec<HuffNode>) {
     if leaf_count == 1 {
         let (symbol, &count) = counts.iter().enumerate().find(|&(_, &c)| c > 0).unwrap();
         nodes.push(HuffNode {
-            count,
-            symbol,
-            parent: None,
+            count: count as u32,
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/huff0/huff0_encoder.rs:1291]: @coderabbitai Fixed, and you were right about the reachability: the narrowing to `u32` is mine, and on a public entry point a histogram that does not fit truncated into a leaf, overflowed at the first merge that crossed the boundary, and at exactly `u32::MAX` produced a leaf indistinguishable from the sentinel that marks an unbuilt node — which the merge loop would then take as a child.

The bound went into `build_huffman_leaf_depths_into` rather than into each entry point, because that is the single place the narrowing happens: both `build_from_counts` and the gated cheap path reach it, so one check covers them by construction and cannot drift apart. It is folded into the pass that already counts the leaves, so it costs one add per non-zero symbol, and the total is accumulated with `checked_add` so it cannot wrap on the way to the comparison. The alphabet bound stays where it was — `counts.len() <= 256` already keeps the `u16` symbol and parent fields in range, since a parent index tops out at `2 * 256 - 1`.

Two regression tests, one per entry point, plus one asserting the largest histogram the encoder can actually produce (a full 128 KiB section) still builds a tree with the frequencies intact. All three fail on the previous commit.

[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/huff0/huff0_encoder.rs:1291]: @coderabbitai Correction to my previous reply: you were right about the placement and I was wrong. I put the bound in the tree builder, arguing that both entry points reach it so one check covers them by construction. The i686 job disproved that within the hour — on a 32-bit target the cheap path's table-log pick sums the counts in a `usize` and overflows on the very input the bound exists to reject, before the builder is ever reached, so the panic arrived from the sum with a different message.

The check now sits at both entry points ahead of anything that reads the histogram, exactly as you described, and the builder keeps a `debug_assert` for the invariant it now holds by construction. The total accumulates in a `u64` so the bound and its message read the same on 32- and 64-bit.

The alphabet half was also right and I dismissed it: only the search path asserted `counts.len() <= 256`, and the cheap path reached the weight buffers and node indices — both sized for 256 symbols — without one. That bound moved to the same place, with a test.

File: zstd/examples/slide_oversized_table.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+//! `hashLog` — which leaves a configuration the advanced API can ask for and
+//! the level presets never produce: a table of a million entries over a window
+//! of a kilobyte, sliding once per kilobyte of input.
+//!
+//! That is the shape where sliding the table's indices and rebuilding it from
+//! the retained bytes cost very different amounts, so it is the fixture for
+//! deciding between them.
+//!
+//! Build: cargo build --profile bench -p structured-zstd
+//!          --example slide_oversized_table --features hash,std,dict-builder
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/examples/slide_oversized_table.rs]: @coderabbitai Fixed in 99e6c818: the build line now reads `cargo build --profile bench -p ffi-bench --example slide_oversized_table`. The old `--features hash,std,dict-builder` went with it, since `ffi-bench` has no `std` feature and the example needs none; the documented command builds as written.

File: zstd/examples/slide_oversized_table.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+fn main() {
+    let args: Vec<String> = env::args().collect();
+    let window_log: u32 = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(10);
+    let hash_log: u32 = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(20);
+    let frame_bytes: usize = args
+        .get(3)
+        .and_then(|s| s.parse().ok())
+        .unwrap_or(4 * 1024 * 1024);
+    let iters: u32 = args.get(4).and_then(|s| s.parse().ok()).unwrap_or(4);
+    let dict_path: Option<&str> = args.get(5).map(|s| s.as_str());
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/examples/slide_oversized_table.rs:59]: @coderabbitai Won't require the dictionary: it would not reach the 2^20 table either. Measured with a raw-content dictionary (Fast, windowLog 10, 4 MiB, one iteration): hashLog 11 and hashLog 20 both print `heap=329984 out=449258`; without one, both print `heap=62768 out=449435`. A dictionary frame runs the dictionary's own table geometry (`resolve_level_params_with_dict`) and never reads the requested hashLog, so no mode builds the oversized shape.

What was wrong is the example's header, which claimed the opposite. Fixed in 99e6c818: it now states that neither mode builds that table, why, and that the `heap=` figure is the check. The dictionary argument stays optional because both arms are meaningful.

Upstream builds a loaded dictionary's CDict from the requested parameters, so there the override does reach the tables. That divergence is in the parameter resolution rather than in this PR, and is tracked in #505, which also covers re-measuring the slide once the shape becomes reachable.
